subsys/bluetooth/services/hids_c.c

```diff
@@ -0,0 +1,851 @@
+{
+	k_mem_slab_free(&bt_gatt_hids_c_reports_mem, (void**)repp);
+	*repp = NULL;
+}
+
+/**
+ * @brief Create new report structure
+ *
+ * Allocate memory for new report and fill it with data from given report
+ * characteristic.
+ *
+ * @param hids_c   HIDS client object.
+ * @param repp     Pointer with would be set to newly created report.
+ * @param dm       Discovery object.
+ * @param rep_chrc Report characteristic.
+ *
+ * @return 0 or negative error code.
+ *
+ * @sa rep_alloc
+ * @sa rep_free
+ *
+ * @note If error is found during report initialization,
+ *       allocated report memory is freed.
+ */
+static int rep_new(struct bt_gatt_hids_c *hids_c,
+		   struct bt_gatt_hids_c_rep_info **repp,
+		   const struct bt_gatt_dm *dm,
+		   const struct bt_gatt_attr *rep_chrc)
+{
+	struct bt_gatt_hids_c_rep_info *rep;
+	const struct bt_gatt_attr *gatt_desc;
+	rep = rep_alloc();
+	if (!rep) {
+		LOG_ERR("No memory to create new report");
+		return -ENOMEM;
+	}
+	struct bt_gatt_chrc *chrc_val = bt_gatt_dm_attr_chrc_val(rep_chrc);
+	assert(chrc_val); /* This is internal function and it has to
+	                   * be called with characteristic attribute */
+	/* Guess the type depending on the flags */
+	if (chrc_val->properties & BT_GATT_CHRC_WRITE_WITHOUT_RESP) {
+		/* Only Out Report has this properties */
+		rep->ref.type = BT_GATT_HIDS_C_REPORT_TYPE_OUTPUT;
+	} else if (chrc_val->properties & BT_GATT_CHRC_NOTIFY){
+		/* Only In Report has this properties */
+		rep->ref.type = BT_GATT_HIDS_C_REPORT_TYPE_INPUT;
+		/* And it has CCC descriptor */
+		gatt_desc = bt_gatt_dm_desc_by_uuid(dm, rep_chrc, BT_UUID_GATT_CCC);
+		if (!gatt_desc) {
+			LOG_ERR("Missing Input Report CCC.");
+			rep_free(&rep);
+			return -EINVAL;
+		}
+		rep->handlers.ccc = gatt_desc->handle;
+	} else {
+		/* It has to be feature report */
+		rep->ref.type = BT_GATT_HIDS_C_REPORT_TYPE_FEATURE;
+	}
+	/* Get common report handlers */
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, rep_chrc, BT_UUID_HIDS_REPORT_REF);
+	if (!gatt_desc) {
+		/* Not required for BOOT report */
+		if (!bt_uuid_cmp(chrc_val->uuid, BT_UUID_HIDS_REPORT)) {
+			LOG_ERR("Missing Report Reference.");
+			rep_free(&rep);
+			return -EINVAL;
+		}
+	} else {
+		rep->handlers.ref = gatt_desc->handle;
+	}
+	gatt_desc = bt_gatt_dm_desc_by_uuid(dm, rep_chrc, chrc_val->uuid);
+	if (!gatt_desc) {
+		LOG_ERR("Missing Report Value.");
+		rep_free(&rep);
+		return -EINVAL;
+	}
+	rep->handlers.val = gatt_desc->handle;
+	/* Remember the parent object */
+	rep->hids_c = hids_c;
+
+	*repp = rep;
+	return 0;
+}
+
+/**
+ * @brief Create new array of pointers to the reports
+ *
+ * Create array of pointers to reports of given number of entities.
+ *
+ * @param cnt Array size.
+ *
+ * @return Pointer to the array of NULL if no memory is available.
+ *
+ * @sa repptr_array_free
+ */
+static struct bt_gatt_hids_c_rep_info **repptr_array_new(size_t cnt)
+{
+	struct bt_gatt_hids_c_rep_info **repp = k_malloc(cnt * sizeof(void *));
```

Comment:
doesn't make big difference but `sizeof(*repp)` would use correct pointer type
the same in `memset`

Reply:
Fixed